File: src/backends/line.rs
use super::common::{Backend, BackendError, SendOption};
use async_trait::async_trait;
use serde::{Deserialize, Serialize};

const API_URL: &str = "https://notify-api.line.me/api/notify";

#[derive(Deserialize, Serialize)]
struct Body {
    message: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LineConfig {
    pub line: LineBackend,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LineBackend {
    token: String,
}

#[async_trait]
impl Backend for LineBackend {
    async fn send(&self, msg: &str, title: &str, _option: &SendOption) -> Result<(), BackendError> {
        let body = &Body {
            message: format!("{}\n{}", title, msg.to_string()),
        };
        let req = match surf::post(API_URL)
            .header("Authorization", format!("Bearer {}", self.token))
            .body_json(body)
        {
            Ok(req) => req,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        let res = match req.await {
            Ok(res) => res,
            Err(err) => return Err(BackendError::Message(err.to_string())),
        };
        if res.status() != 200 {
            return Err(BackendError::Message(res.status().to_string()));
        }

        Ok(())
    }
}

File: backends/line.go
package backends

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

func init() {
	backends["line"] = NewLine()
}

const LINE_API_URL = "https://notify-api.line.me/api/notify"

type LineConfig struct {
	Token string `mapstructure:"token" validate:"required"`
}

type Line struct {
}

func NewLine() BackendInterface {
	return &Line{}
}

func (*Line) GetConfig() interface{} {
	return LineConfig{}
}

func (*Line) Send(configIface interface{}, title string, message string, status *bool) error {
	config, ok := configIface.(LineConfig)
	if !ok {
		return fmt.Errorf("invalid config")
	}
	form := url.Values{}
	form.Add("message", fmt.Sprintf("%s\n%s", title, message))
	body := strings.NewReader(form.Encode())
	req, err := http.NewRequest("POST", LINE_API_URL, body)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", config.Token))
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		return fmt.Errorf("line: %s", res.Status)
	}

	return nil
}
